### PR TITLE
Determine bzero based on absolute s/mm^2

### LINKED ISSUE #136: [BUG] bval=5 is interpreted as a non zero gradient
The issue has been discovered with HCP data that have bvalues:

```
5       5       1490     1495       1500      1495     1490       ...
```

As a result, the program throws an exception and fails:

> No zero gradients!
terminate called without an active exception

I believe [this](https://github.com/pnlbwh/ukftractography/blob/4a4aba614f6337041e24c26522da0031d0132f6b/ukf/dwi_normalize.cc#L68) is the problematic line that treats gradients > 0.05 as non zero gradients:

> nonZeroGradientFlag.push_back( gradient_magnitude > 0.05F );

The threshold should be set at 50 as we do in all other cases--harmonization, TBSS etc.

---

Hi @pieper , is my interpretation of the `F` identifier i.e. `0.05F` correct?

### PR DESCRIPTION
**Pull request by tashrifbillah**
Fixes #136

### CODE REVIEW

**Review by pieper — APPROVED** [review]
LGTM 👍

File: ukf/dwi_normalize.cc [hunk 1]
```diff
@@ -64,8 +75,12 @@ void dwiNormalize(const Nrrd *raw, Nrrd *& normalized)
         std::cout << "The gradients must have 3 components!" << std::endl;
         throw;
         }
+
       const float gradient_magnitude = sqrt(gx * gx + gy * gy + gz * gz);
-      nonZeroGradientFlag.push_back( gradient_magnitude > 0.05F );
+
+      // See https://github.com/pnlbwh/ukftractography/issues/136#issuecomment-753173654
+      // for details about the following criteria
+      nonZeroGradientFlag.push_back( bmax * gradient_magnitude * gradient_magnitude > 50 );
```
Comment (tashrifbillah) [inline ukf/dwi_normalize.cc:83]: Hi @pieper , I need a little help. The above condition evaluates correctly in my workstation but not in ERIS cluster. I wonder if it has anything to do with GCC version incompatibility between the machines. Here is a concrete example:

> bmax * gradient_magnitude * gradient_magnitude

`1000 * 0.447214 * 0.447214` evaluates to `3.8` which should be `200`. I see that I read bmax as `%d` so I tried the following typecast but to no avail:

> (float)bmax * gradient_magnitude * gradient_magnitude

Bottom line, the modification broke UKF :( Can you guide us?

cc @yrathi

Comment (tashrifbillah) [reply ukf/dwi_normalize.cc:83]: Hi @pieper ,

I have better information about the issue:

* a `char` pointer is declared [here](https://github.com/pnlbwh/ukftractography/blob/master/ukf/dwi_normalize.cc#L56)
* after bmax is read for [the first time](https://github.com/pnlbwh/ukftractography/blob/master/ukf/dwi_normalize.cc#L64), its value changes in subsequent iterations of the `for` loop (!!!):
```python
  for( unsigned int i = 0; i < nrrdKeyValueSize(raw); i++ )
    {
    char *key;
    char *value;
    nrrdKeyValueIndex(raw, &key, &value, i);
    std::string keyStr(key);

    // Obtain DWMRI_b-value
    int bmax;
    if( keyStr.length() == 13 && !keyStr.compare("DWMRI_b-value") )
      {
        sscanf(value, "%d", &bmax);
        std::cout << "DWMRI_b-value " << bmax << std::endl;

      }


    if( keyStr.length() > 14 && !keyStr.substr(0, 14).compare("DWMRI_gradient") )
      {
      float gx, gy, gz;
      if( 3 != sscanf(value, "%50f %50f %50f", &gx, &gy, &gz) )
        {
        std::cout << "The gradients must have 3 components!" << std::endl;
        throw;
        }
```

As you can see, that `value` pointer is used in a second `sscanf`. Do you think the second reading has an effect on the value stored at `&bmax`?

I am continuing to dig in though ...

Comment (tashrifbillah) [reply ukf/dwi_normalize.cc:83]: Okay, it does seem that the second use of `value` pointer with `sscanf` caused this issue. If I make the following changes, the issue gets fixed:

```python
    char *value;
    char *value1;
    nrrdKeyValueIndex(raw, &key, &value, i);
    nrrdKeyValueIndex(raw, &key, &value1, i);
    ...
    ...
    sscanf(value1, "%d", &bmax);
```

Now I am exploring how we can do it using only one `value` pointer.

Comment (tashrifbillah) [reply ukf/dwi_normalize.cc:83]: Okay, I seem to have fixed it ... PR to follow through ...

Comment (pieper) [reply ukf/dwi_normalize.cc:83]: Nice work @tashrifbillah 👍   working with raw C pointers can be like juggling scalpels sometimes.  Best to use extreme care.

Comment (tashrifbillah) [reply ukf/dwi_normalize.cc:83]: Hi Steve, thank you for your encouragement. Working with these unforeseen issues, I am starting to love UKF maintenance ;)

Here is the revised PR #138

**Pull request merged**